### ISSUE TITLE
Lower load_memory_map memory consumption

### ISSUE THREAD

**Issue by callegarimattia** [weareprestatech/hotpdf#15]
Right now to load the Bible (an example of a big [pdf file](https://github.com/weareprestatech/hotpdf/files/13933140/The-Holy-Bible-King-James-Version.pdf) of ~700pages) the memory usage skyrockets to around 1.3 GiB.

The big memory allocation is of course line 108 in  [memory_map.py](https://github.com/weareprestatech/hotpdf/blob/main/hotpdf/memory_map.py)
```python
char_x0, char_y0, char_x1, _ = [float(char_coord) for char_coord in char_bbox.split()]
```
Here's the memray flamegraph showing the allocation call:
![image](https://github.com/weareprestatech/hotpdf/assets/106533898/1821d43e-0451-4344-baf5-6b3dc8b7480e)

Any ideas? 😅
@krishnasism @aptakhin

**Comment by krishnasism** [comment]
With sparse matrix we saved a lot of memory anyway. Also going iterparse saved another 30% 

I think now it's ok to have 1.3 gb (ok not that great) for the Bible, because we were somewhere around 4-8GB with a simple 100page file. 

One of the main reasons to write hotpdf was also to prevent our workers from going OOM. For example the 100-page file took >4GB before and killed out workers. Given the pages that we parse on average, our workers won't go OOM. We went from 4GB per file down to less than 300MB. 

We can definitely take a look at how to reduce this even more. But with the multiple data structures and the nature of handling strings, I doubt we'll see a significant improvements. 

In any case let's aim for <1GB usage for the Bible. I'll take a look at redundant data storage in memory.

**Comment by callegarimattia** [comment]
Makes sense.
To be fair I think the biggest bottleneck is python itself.
May make sense to have the sparse_matrix and trie implementation in a different language to handle better the memory.
What do you think?

**Comment by krishnasism** [comment]
Yep! Next we can try and experimenting with writing the underlying data structures in Rust. 

We can take a look at PyO3 + maturin

**Comment by callegarimattia** [comment]
Let's do it!

**Comment by callegarimattia** [comment]
Linking #34

**Comment by krishnasism** [comment]
Still, a couple of more PRs to go before we can close this 😉 
Will take a look at more ways to optimised, also feel free to point out any inefficient code that you find, we can rewrite them
